ee/control/notify_consumer.go

```diff
@@ -0,0 +1,167 @@
+package control
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	desktopRunner "github.com/kolide/launcher/ee/desktop/runner"
+	"github.com/kolide/launcher/pkg/osquery"
+	"go.etcd.io/bbolt"
+)
+
+// Consumes notifications from control server, tracks when notifications are sent to end user
+type NotificationConsumer struct {
+	db              *bbolt.DB
+	runner          notifier
+	logger          log.Logger
+	notificationTtl time.Duration // How long until a notification expires -- i.e. how long until we can re-send that identical notification
+}
+
+// The desktop runner fullfils this interface -- it exists primarily for testing purposes.
+type notifier interface {
+	SendNotification(title, body string) error
+}
+
+// Represents notification received from control server; SentAt is set by this consumer after sending.
+type notification struct {
+	Title  string    `json:"title"`
+	Body   string    `json:"body"`
+	UUID   string    `json:"uuid"`
+	SentAt time.Time `json:"sent_at,omitempty"`
+}
+
+// Identifier for this consumer.
+const NotificationSubsystem = "desktop_notifier"
+
+type notificationConsumerOption func(*NotificationConsumer)
+
+func WithLogger(logger log.Logger) notificationConsumerOption {
+	return func(nc *NotificationConsumer) {
+		nc.logger = log.With(logger,
+			"component", NotificationSubsystem,
+		)
+	}
+}
+
+func WithNotificationTtl(ttl time.Duration) notificationConsumerOption {
+	return func(nc *NotificationConsumer) {
+		nc.notificationTtl = ttl
+	}
+}
+
+func NewNotifyConsumer(db *bbolt.DB, runner *desktopRunner.DesktopUsersProcessesRunner, opts ...notificationConsumerOption) (*NotificationConsumer, error) {
+	// Create the bucket to track sent notifications if it doesn't exist
+	if err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte(osquery.SentNotificationsBucket))
+		if err != nil {
+			return fmt.Errorf("could not create bucket: %w", err)
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("cannot create notifier without db bucket: %w", err)
+	}
+
+	nc := &NotificationConsumer{
+		db:              db,
+		runner:          runner,
+		logger:          log.NewNopLogger(),
+		notificationTtl: time.Hour * 1,
+	}
+
+	for _, opt := range opts {
+		opt(nc)
+	}
+
+	return nc, nil
+}
+
+func (nc *NotificationConsumer) Update(data io.Reader) error {
+	var notificationsToProcess []notification
+	if err := json.NewDecoder(data).Decode(&notificationsToProcess); err != nil {
+		return fmt.Errorf("failed to decode notification data: %w", err)
+	}
+
+	for _, notificationToProcess := range notificationsToProcess {
+		nc.notify(notificationToProcess)
+	}
+
```

Comment:
maybe should consider adding some delete logic, unless it's already there and I missed it ... keep only the last 100 notifications or something? or perhaps delete any that are more than 24 hours old. Perhaps save that for next PR?

Reply:
You didn't miss it -- there's no deletion logic right now. I was thinking we'd probably want to retain the notifications for at least some amount of time, but wasn't sure what was reasonable/what kind of records we'd want to keep about notifications. I imagine we wouldn't want to be pinging the end user more than once or twice a day, so maybe retaining for a week would be fine?

I can save for the next PR while we discuss, since this one is on the larger size anyway!

Reply:
There's cleanup logic now